STM32F405RGTx/Core/Src/rtos_threads/state_machine.cpp

```diff
@@ -0,0 +1,440 @@
+        // error code
+        return SevereDangerFault;
+    }
+
+    // DC fault checking
+    if (global_mc_data.dc_voltage > MAX_DCVOLTAGE_SEVERE) {
+        // error code
+        return SevereDangerFault;
+    }
+    else if (global_mc_data.dc_voltage < MIN_DCVOLTAGE_SEVERE) {
+        // error code
+        return SevereDangerFault;
+    }
+    if (global_mc_data.dc_cap_temp > MAX_DCCAP_TEMP_SEVERE) {
+        return SevereDangerFault;
+    }
+    return NoFault;
+}
+
+State_t StateMachineThread::InitializeEvent(void) {
+    return Idle;
+}
+
+State_t StateMachineThread::IdleEvent(void) {
+    SetLedColour(0.0, 50.0, 0.0);
+    State_t normal_fault_check = NormalFaultChecking();
+    State_t severe_fault_check = SevereFaultChecking();
+    if (severe_fault_check == SevereDangerFault) {
+        return severe_fault_check;
+    }
+    else if (normal_fault_check == NormalDangerFault) {
+        return normal_fault_check;
+    }
+
+    // TODO: Make list of the threads that need to be turned on or turned off
+        // On: Measurements Thread, CANZThread
+        // Off: PIDThread, VHzThread, SVPWMThread
+    MeasurementsThread::resumeMeasurements();
+
+    if (!Queue_empty(&RX_QUEUE)) {
+        CANFrame rx_frame = CANBus_get_frame();
+        uint8_t state_id = CANFrame_get_field(&rx_frame, STATE_ID);
+        idle_state_id = state_id;
+        if (state_id == AUTO_PILOT) {
+            return AutoPilot;
+        }
+        else if (state_id == MANUAL_OPERATION_WAITING) {
+            return ManualControl;
+        } 
+    }
+
+    return Idle;
+}
+
+State_t StateMachineThread::AutoPilotEvent(void)
+{
+    // Set LED colour to yellow
+    SetLedColour(50.0, 50.0, 0.0);
+
+    // Send ACK on CAN when stop complete 
+    if (NewState != CurrentState) {
+        CANFrame tx_frame = CANFrame_init(MOTOR_CONTROLLER_STATE_CHANGE_ACK_NACK);
+        CANFrame_set_field(&tx_frame, MOTOR_CONTROLLER_STATE_ID_ACK_NACK, idle_state_id);
+        if (CANBus_put_frame(&tx_frame) != HAL_OK) { Error_Handler(); }
+    }
+
+    // TODO: Make list of the threads that need to be turned on or turned off
+        // On: Measurements Thread, CANThread
+        // Off: PIDThread, VHzThread, SVPWMThread
+    MeasurementsThread::resumeMeasurements();
+
+    // Receive CAN frame
+    if (!Queue_empty(&RX_QUEUE)) {
+        CANFrame rx_frame = CANBus_get_frame();
+        uint8_t state_id = CANFrame_get_field(&rx_frame, STATE_ID);
+        int distance_to_end = TRACK_LENGTH - global_mc_data.curr_pos;
+        if (state_id == EMERGENCY_BRAKE || state_id == SYSTEM_FAILURE) {
+            return SevereDangerFault;
+            // TODO: Talk to Ryan/Dev when to enter MinorDangerFault
+        }
+        else if (state_id == BRAKING || distance_to_end <= DISTANCE_THRESHOLD) {
+            return Idle;
+        } 
+    }
+    return AutoPilot;
+}
+
+State_t StateMachineThread::ManualControlEvent(void)
```

Comment:
So for this manual control mode, the MC will be listening on the CAN bus for which speed to set the `target_speed` to. I don't know which message this is sent in, but it SHOULD be in `config.h` of the CAN library.

Reply:
Do you mean curr_speed or is there a way to calculate the target_speed?